File: ios/Runner/MetadataWriteHandler.swift
import Flutter
import Photos
import AVFoundation
import CoreLocation
import ImageIO
import MobileCoreServices

/// Handles writing dive metadata to photos and videos via platform channel.
///
/// Supports:
/// - JPEG/HEIC/HEIF photos via CGImageDestination
/// - MOV/MP4 videos via AVFoundation
class MetadataWriteHandler: NSObject {
    private let channel: FlutterMethodChannel

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "com.submersion.app/metadata",
            binaryMessenger: messenger
        )
        super.init()
        channel.setMethodCallHandler(handle)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "writeMetadata":
            guard let args = call.arguments as? [String: Any],
                  let assetId = args["assetId"] as? String,
                  let metadata = args["metadata"] as? [String: Any],
                  let description = args["description"] as? String,
                  let isVideo = args["isVideo"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }

            let keepOriginal = args["keepOriginal"] as? Bool ?? false

            writeMetadata(
                assetId: assetId,
                metadata: metadata,
                description: description,
                isVideo: isVideo,
                keepOriginal: keepOriginal,
                result: result
            )

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func writeMetadata(
        assetId: String,
        metadata: [String: Any],
        description: String,
        isVideo: Bool,
        keepOriginal: Bool,
        result: @escaping FlutterResult
    ) {
        // Check photo library authorization
        let status = PHPhotoLibrary.authorizationStatus(for: .readWrite)
        guard status == .authorized || status == .limited else {
            result(FlutterError(
                code: "PERMISSION_DENIED",
                message: "Photo library access not authorized",
                details: nil
            ))
            return
        }

        // Find the asset
        let fetchResult = PHAsset.fetchAssets(withLocalIdentifiers: [assetId], options: nil)
        guard let asset = fetchResult.firstObject else {
            result(FlutterError(
                code: "ASSET_NOT_FOUND",
                message: "Asset not found with ID: \(assetId)",
                details: nil
            ))
            return
        }

        // Check if asset is editable
        guard asset.canPerform(.properties) else {
            result(FlutterError(
                code: "READ_ONLY",
                message: "Asset cannot be modified (may be in iCloud or shared album)",
                details: nil
            ))
            return
        }

        if isVideo {
            writeVideoMetadata(asset: asset, metadata: metadata, description: description, keepOriginal: keepOriginal, result: result)
        } else {
            writePhotoMetadata(asset: asset, metadata: metadata, description: description, result: result)
        }
    }

    // MARK: - Photo Metadata

    private func writePhotoMetadata(
        asset: PHAsset,
        metadata: [String: Any],
        description: String,
        result: @escaping FlutterResult
    ) {
        let options = PHContentEditingInputRequestOptions()
        options.isNetworkAccessAllowed = true

        asset.requestContentEditingInput(with: options) { [weak self] input, info in
            guard let self = self, let input = input else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Could not get content editing input",
                        details: nil
                    ))
                }
                return
            }

            guard let inputURL = input.fullSizeImageURL else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Could not get image URL",
                        details: nil
                    ))
                }
                return
            }

            // Read existing image data
            guard let imageSource = CGImageSourceCreateWithURL(inputURL as CFURL, nil) else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Could not read image source",
                        details: nil
                    ))
                }
                return
            }

            // Get existing metadata
            var existingMetadata = CGImageSourceCopyPropertiesAtIndex(imageSource, 0, nil) as? [String: Any] ?? [:]

            // Build GPS metadata
            var gpsDict = existingMetadata[kCGImagePropertyGPSDictionary as String] as? [String: Any] ?? [:]

            if let depth = metadata["depthMeters"] as? Double {
                gpsDict[kCGImagePropertyGPSAltitude as String] = abs(depth)
                gpsDict[kCGImagePropertyGPSAltitudeRef as String] = 1 // Below sea level
            }

            if let lat = metadata["latitude"] as? Double,
               let lon = metadata["longitude"] as? Double {
                gpsDict[kCGImagePropertyGPSLatitude as String] = abs(lat)
                gpsDict[kCGImagePropertyGPSLatitudeRef as String] = lat >= 0 ? "N" : "S"
                gpsDict[kCGImagePropertyGPSLongitude as String] = abs(lon)
                gpsDict[kCGImagePropertyGPSLongitudeRef as String] = lon >= 0 ? "E" : "W"
            }

            existingMetadata[kCGImagePropertyGPSDictionary as String] = gpsDict

            // Build TIFF metadata for description
            var tiffDict = existingMetadata[kCGImagePropertyTIFFDictionary as String] as? [String: Any] ?? [:]
            tiffDict[kCGImagePropertyTIFFImageDescription as String] = description
            existingMetadata[kCGImagePropertyTIFFDictionary as String] = tiffDict

            // Build EXIF metadata
            var exifDict = existingMetadata[kCGImagePropertyExifDictionary as String] as? [String: Any] ?? [:]
            exifDict[kCGImagePropertyExifUserComment as String] = description
            existingMetadata[kCGImagePropertyExifDictionary as String] = exifDict

            // Create output
            let output = PHContentEditingOutput(contentEditingInput: input)
            let outputURL = output.renderedContentURL

            // Determine UTI based on original file
            let uti = self.getImageUTI(for: inputURL)

            guard let destination = CGImageDestinationCreateWithURL(
                outputURL as CFURL,
                uti,
                1,
                nil
            ) else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Could not create image destination",
                        details: nil
                    ))
                }
                return
            }

            // Copy the image with updated metadata
            CGImageDestinationAddImageFromSource(destination, imageSource, 0, existingMetadata as CFDictionary)

            guard CGImageDestinationFinalize(destination) else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Could not finalize image with metadata",
                        details: nil
                    ))
                }
                return
            }

            // Set adjustment data (required for Photos to accept the edit)
            let adjustmentData = PHAdjustmentData(
                formatIdentifier: "com.submersion.app.metadata",
                formatVersion: "1.0",
                data: description.data(using: .utf8) ?? Data()
            )
            output.adjustmentData = adjustmentData

            // Commit the changes
            PHPhotoLibrary.shared().performChanges({
                let request = PHAssetChangeRequest(for: asset)
                request.contentEditingOutput = output
            }) { success, error in
                DispatchQueue.main.async {
                    if success {
                        result(true)
                    } else {
                        result(FlutterError(
                            code: "WRITE_FAILED",
                            message: error?.localizedDescription ?? "Failed to save changes",
                            details: nil
                        ))
                    }
                }
            }
        }
    }

    private func getImageUTI(for url: URL) -> CFString {
        let ext = url.pathExtension.lowercased()
        switch ext {
        case "heic", "heif":
            return AVFileType.heic as CFString
        case "png":
            return kUTTypePNG
        case "jpg", "jpeg":
            return kUTTypeJPEG
        default:
            return kUTTypeJPEG
        }
    }

    // MARK: - Video Metadata

    private func writeVideoMetadata(
        asset: PHAsset,
        metadata: [String: Any],
        description: String,
        keepOriginal: Bool,
        result: @escaping FlutterResult
    ) {
        let options = PHVideoRequestOptions()
        options.isNetworkAccessAllowed = true
        options.version = .current

        PHImageManager.default().requestAVAsset(forVideo: asset, options: options) { [weak self] avAsset, audioMix, info in
            guard let self = self, let urlAsset = avAsset as? AVURLAsset else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Could not get video asset",
                        details: nil
                    ))
                }
                return
            }

            self.writeVideoMetadataToAsset(
                asset: asset,
                videoURL: urlAsset.url,
                metadata: metadata,
                description: description,
                keepOriginal: keepOriginal,
                result: result
            )
        }
    }

    private func writeVideoMetadataToAsset(
        asset: PHAsset,
        videoURL: URL,
        metadata: [String: Any],
        description: String,
        keepOriginal: Bool,
        result: @escaping FlutterResult
    ) {
        // Create AVAsset for reading
        let avAsset = AVAsset(url: videoURL)

        // Create export session
        guard let exportSession = AVAssetExportSession(asset: avAsset, presetName: AVAssetExportPresetPassthrough) else {
            DispatchQueue.main.async {
                result(FlutterError(
                    code: "WRITE_FAILED",
                    message: "Could not create export session",
                    details: nil
                ))
            }
            return
        }

        // Create temporary output URL
        let tempDir = FileManager.default.temporaryDirectory
        let outputURL = tempDir.appendingPathComponent(UUID().uuidString + ".mov")

        exportSession.outputURL = outputURL
        exportSession.outputFileType = .mov

        // Create metadata items
        var metadataItems: [AVMutableMetadataItem] = []

        // Add description
        let descItem = AVMutableMetadataItem()
        descItem.key = AVMetadataKey.commonKeyDescription as NSString
        descItem.keySpace = .common
        descItem.value = description as NSString
        metadataItems.append(descItem)

        // Add location if available
        if let lat = metadata["latitude"] as? Double,
           let lon = metadata["longitude"] as? Double {
            let locationItem = AVMutableMetadataItem()
            locationItem.key = AVMetadataKey.commonKeyLocation as NSString
            locationItem.keySpace = .common
            locationItem.value = "\(lat >= 0 ? "+" : "")\(lat)\(lon >= 0 ? "+" : "")\(lon)/" as NSString
            metadataItems.append(locationItem)
        }

        // Add title with site name
        if let siteName = metadata["siteName"] as? String {
            let titleItem = AVMutableMetadataItem()
            titleItem.key = AVMetadataKey.commonKeyTitle as NSString
            titleItem.keySpace = .common
            titleItem.value = siteName as NSString
            metadataItems.append(titleItem)
        }

        exportSession.metadata = metadataItems

        exportSession.exportAsynchronously {
            switch exportSession.status {
            case .completed:
                // Import the modified video back to Photos
                PHPhotoLibrary.shared().performChanges({
                    let creationRequest = PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: outputURL)
                    creationRequest?.creationDate = asset.creationDate
                    creationRequest?.location = asset.location
                }) { success, error in
                    // Clean up temp file
                    try? FileManager.default.removeItem(at: outputURL)

                    if success {
                        // Delete original if user chose not to keep it
                        if !keepOriginal {
                            PHPhotoLibrary.shared().performChanges({
                                PHAssetChangeRequest.deleteAssets([asset] as NSArray)
                            }) { _, deleteError in
                                // Log but don't fail the operation if delete fails
                                if let deleteError = deleteError {
                                    print("Warning: Could not delete original video: \(deleteError.localizedDescription)")
                                }
                                DispatchQueue.main.async {
                                    result(true)
                                }
                            }
                        } else {
                            DispatchQueue.main.async {
                                result(true)
                            }
                        }
                    } else {
                        DispatchQueue.main.async {
                            result(FlutterError(
                                code: "WRITE_FAILED",
                                message: error?.localizedDescription ?? "Failed to save video",
                                details: nil
                            ))
                        }
                    }
                }

            case .failed:
                try? FileManager.default.removeItem(at: outputURL)
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: exportSession.error?.localizedDescription ?? "Export failed",
                        details: nil
                    ))
                }

            case .cancelled:
                try? FileManager.default.removeItem(at: outputURL)
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "WRITE_FAILED",
                        message: "Export was cancelled",
                        details: nil
                    ))
                }

            default:
                break
            }
        }
    }
}

File: android/app/src/main/kotlin/app/submersion/MetadataWriteHandler.kt
package app.submersion

import android.content.ContentResolver
import android.content.ContentUris
import android.content.ContentValues
import android.content.Context
import android.media.MediaMetadataRetriever
import android.media.MediaMuxer
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import androidx.exifinterface.media.ExifInterface
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.UUID

/**
 * Handles writing dive metadata to photos and videos via platform channel (Android).
 *
 * Supports:
 * - JPEG/PNG photos via ExifInterface
 * - MP4/MOV videos via MediaMuxer (creates new file with metadata)
 */
class MetadataWriteHandler(
    context: Context,
    messenger: BinaryMessenger
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "com.submersion.app/metadata")
    private val appContext = context.applicationContext

    init {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "writeMetadata" -> {
                val assetId = call.argument<String>("assetId")
                val metadata = call.argument<Map<String, Any>>("metadata")
                val description = call.argument<String>("description")
                val isVideo = call.argument<Boolean>("isVideo")
                val keepOriginal = call.argument<Boolean>("keepOriginal") ?: false

                if (assetId == null || metadata == null || description == null || isVideo == null) {
                    result.error("INVALID_ARGS", "Missing required arguments", null)
                    return
                }

                writeMetadata(assetId, metadata, description, isVideo, keepOriginal, result)
            }
            else -> result.notImplemented()
        }
    }

    private fun writeMetadata(
        assetId: String,
        metadata: Map<String, Any>,
        description: String,
        isVideo: Boolean,
        keepOriginal: Boolean,
        result: MethodChannel.Result
    ) {
        try {
            // Parse the asset ID to get the MediaStore URI
            val mediaId = assetId.toLongOrNull()
            if (mediaId == null) {
                result.error("INVALID_ARGS", "Invalid asset ID format: $assetId", null)
                return
            }

            val contentUri = if (isVideo) {
                ContentUris.withAppendedId(MediaStore.Video.Media.EXTERNAL_CONTENT_URI, mediaId)
            } else {
                ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, mediaId)
            }

            if (isVideo) {
                writeVideoMetadata(contentUri, mediaId, metadata, description, keepOriginal, result)
            } else {
                writePhotoMetadata(contentUri, metadata, description, result)
            }
        } catch (e: Exception) {
            result.error("WRITE_FAILED", "Failed to write metadata: ${e.message}", null)
        }
    }

    // MARK: - Photo Metadata

    private fun writePhotoMetadata(
        contentUri: Uri,
        metadata: Map<String, Any>,
        description: String,
        result: MethodChannel.Result
    ) {
        try {
            val resolver = appContext.contentResolver

            // Open the file for writing EXIF data
            resolver.openFileDescriptor(contentUri, "rw")?.use { pfd ->
                val exif = ExifInterface(pfd.fileDescriptor)

                // Write GPS altitude (depth as negative altitude)
                metadata["depthMeters"]?.let { depth ->
                    val depthValue = (depth as Number).toDouble()
                    // GPS altitude is stored as positive value with altitude ref indicating below/above sea level
                    exif.setAltitude(-depthValue) // Negative = below sea level
                }

                // Write GPS coordinates
                val lat = metadata["latitude"] as? Number
                val lon = metadata["longitude"] as? Number
                if (lat != null && lon != null) {
                    exif.setLatLong(lat.toDouble(), lon.toDouble())
                }

                // Write description as ImageDescription and UserComment
                exif.setAttribute(ExifInterface.TAG_IMAGE_DESCRIPTION, description)
                exif.setAttribute(ExifInterface.TAG_USER_COMMENT, description)

                // Save the changes
                exif.saveAttributes()
            } ?: run {
                result.error("WRITE_FAILED", "Could not open file for writing", null)
                return
            }

            result.success(true)
        } catch (e: SecurityException) {
            result.error("PERMISSION_DENIED", "Storage permission denied", null)
        } catch (e: Exception) {
            result.error("WRITE_FAILED", "Failed to write EXIF: ${e.message}", null)
        }
    }

    // MARK: - Video Metadata

    private fun writeVideoMetadata(
        contentUri: Uri,
        mediaId: Long,
        metadata: Map<String, Any>,
        description: String,
        keepOriginal: Boolean,
        result: MethodChannel.Result
    ) {
        try {
            val resolver = appContext.contentResolver

            // Get the original file path
            val projection = arrayOf(MediaStore.Video.Media.DATA, MediaStore.Video.Media.DISPLAY_NAME)
            val cursor = resolver.query(contentUri, projection, null, null, null)

            if (cursor == null || !cursor.moveToFirst()) {
                result.error("ASSET_NOT_FOUND", "Video not found", null)
                cursor?.close()
                return
            }

            val dataIndex = cursor.getColumnIndex(MediaStore.Video.Media.DATA)
            val nameIndex = cursor.getColumnIndex(MediaStore.Video.Media.DISPLAY_NAME)
            val originalPath = cursor.getString(dataIndex)
            val displayName = cursor.getString(nameIndex)
            cursor.close()

            // Create a temporary output file
            val tempDir = appContext.cacheDir
            val outputFile = File(tempDir, "temp_${UUID.randomUUID()}.mp4")

            // Copy video with metadata using MediaMuxer
            val success = copyVideoWithMetadata(
                originalPath,
                outputFile.absolutePath,
                metadata,
                description
            )

            if (!success) {
                outputFile.delete()
                result.error("WRITE_FAILED", "Failed to add metadata to video", null)
                return
            }

            // Insert the new video into MediaStore
            val newVideoUri = insertVideoToMediaStore(
                resolver,
                outputFile,
                displayName,
                description,
                metadata
            )

            // Clean up temp file
            outputFile.delete()

            if (newVideoUri == null) {
                result.error("WRITE_FAILED", "Failed to save new video to library", null)
                return
            }

            // Delete original if user chose not to keep it
            if (!keepOriginal) {
                try {
                    resolver.delete(contentUri, null, null)
                } catch (e: Exception) {
                    // Log but don't fail if delete doesn't work
                    android.util.Log.w("MetadataWriteHandler", "Could not delete original: ${e.message}")
                }
            }

            result.success(true)
        } catch (e: SecurityException) {
            result.error("PERMISSION_DENIED", "Storage permission denied", null)
        } catch (e: Exception) {
            result.error("WRITE_FAILED", "Failed to write video metadata: ${e.message}", null)
        }
    }

    private fun copyVideoWithMetadata(
        inputPath: String,
        outputPath: String,
        metadata: Map<String, Any>,
        description: String
    ): Boolean {
        // For Android, we can't easily add custom metadata to MP4 files without re-encoding.
        // The best approach is to use MediaMuxer to remux the video, but it has limitations.
        // For now, we'll copy the file and update what we can via MediaStore.

        // Simple file copy (MediaStore metadata will be set separately)
        return try {
            FileInputStream(inputPath).use { input ->
                FileOutputStream(outputPath).use { output ->
                    input.copyTo(output)
                }
            }
            true
        } catch (e: Exception) {
            android.util.Log.e("MetadataWriteHandler", "Failed to copy video: ${e.message}")
            false
        }
    }

    private fun insertVideoToMediaStore(
        resolver: ContentResolver,
        videoFile: File,
        displayName: String,
        description: String,
        metadata: Map<String, Any>
    ): Uri? {
        val values = ContentValues().apply {
            put(MediaStore.Video.Media.DISPLAY_NAME, displayName)
            put(MediaStore.Video.Media.MIME_TYPE, "video/mp4")
            put(MediaStore.Video.Media.DESCRIPTION, description)

            // Add GPS coordinates if available
            val lat = metadata["latitude"] as? Number
            val lon = metadata["longitude"] as? Number
            if (lat != null && lon != null) {
                put(MediaStore.Video.Media.LATITUDE, lat.toDouble())
                put(MediaStore.Video.Media.LONGITUDE, lon.toDouble())
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                put(MediaStore.Video.Media.RELATIVE_PATH, Environment.DIRECTORY_MOVIES)
                put(MediaStore.Video.Media.IS_PENDING, 1)
            }
        }

        val uri = resolver.insert(MediaStore.Video.Media.EXTERNAL_CONTENT_URI, values)
            ?: return null

        try {
            // Copy file content to the new MediaStore entry
            resolver.openOutputStream(uri)?.use { output ->
                FileInputStream(videoFile).use { input ->
                    input.copyTo(output)
                }
            }

            // Mark as no longer pending (Android Q+)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                val updateValues = ContentValues().apply {
                    put(MediaStore.Video.Media.IS_PENDING, 0)
                }
                resolver.update(uri, updateValues, null, null)
            }

            return uri
        } catch (e: Exception) {
            // Clean up on failure
            resolver.delete(uri, null, null)
            throw e
        }
    }
}
